src/elements/Input/Input.js

```diff
@@ -1,118 +1,231 @@
+    'ui',
+    size,
+    useValueAndKey(actionPosition, 'action') || useKeyOnly(action, 'action'),
+    useKeyOnly(disabled, 'disabled'),
+    useKeyOnly(error, 'error'),
+    useKeyOnly(focus, 'focus'),
+    useKeyOnly(fluid, 'fluid'),
+    useKeyOnly(inverted, 'inverted'),
+    useValueAndKey(labelPosition, 'labeled') || useKeyOnly(label, 'labeled'),
+    useKeyOnly(loading, 'loading'),
+    useKeyOnly(transparent, 'transparent'),
+    useValueAndKey(iconPosition, 'icon') || useKeyOnly(icon, 'icon'),
+    className,
+    'input',
+  )
 
-  static _meta = {
-    name: 'Input',
-    type: META.TYPES.ELEMENT,
-  }
+  const rest = getUnhandledProps(Input, props)
+  const ElementType = getElementType(Input, props)
+  const inputProps = _.pick(props, htmlInputPropNames)
 
-  render() {
-    const { className, children, icon, input, type } = this.props
-    // Semantic supports actions and labels on either side of an input.
-    // The element must be on the same side as the indicated class.
-    // We first determine the left/right classes for each type of child,
-    //   then we extract the children and place them on the correct side
-    //   of the input.
-    const isLeftAction = _.includes(className, 'left action')
-    const isRightAction = !isLeftAction && _.includes(className, 'action')
-    const isRightLabeled = _.includes(className, 'right labeled')
-    const isLeftLabeled = !isRightLabeled && _.includes(className, 'labeled')
-
-    const labelChildren = []
-    const actionChildren = []
-
-    Children.forEach(children, child => {
-      const isAction = _.includes(['Button', 'Dropdown', 'Select'], child.type._meta.name)
-      const isLabel = child.type._meta.name === 'Label'
-
-      if (isAction) {
-        actionChildren.push(child)
-      } else if (isLabel) {
-        labelChildren.push(child)
-      }
-    })
-
-    const classes = classNames(
-      'ui',
-      className,
-      'input'
-    )
-    const unhandledProps = getUnhandledProps(Input, this.props)
-    const inputProps = _.pick(unhandledProps, inputPropNames)
-    const rest = _.omit(unhandledProps, inputPropNames)
-    const ElementType = getElementType(Input, this.props)
-    return (
-      <ElementType {...rest} className={classes}>
-        {isLeftLabeled && labelChildren}
-        {isLeftAction && actionChildren}
-        <input {...inputProps} {...input} type={type} />
-        {Icon.create(icon)}
-        {isRightLabeled && labelChildren}
-        {isRightAction && actionChildren}
-      </ElementType>
-    )
+  if (children) {
+    return <ElementType {...rest} className={classes}>{children}</ElementType>
   }
+
+  const actionElement = Button.create(action, elProps => ({
+    className: cx(
+      // all action components should have the button className
+      !_.includes(elProps.className, 'button') && 'button',
+    ),
+  }))
+  const iconElement = Icon.create(icon)
+  const labelElement = Label.create(label, elProps => ({
+    className: cx(
```

Comment:
In early tests, the object form was 8-10x slower than the short circuited argument approach.  This is one reason we went with the lib utils we have for className buildup.  The timing difference will probably never be an issue.  I just continue it out of consistency and habit.